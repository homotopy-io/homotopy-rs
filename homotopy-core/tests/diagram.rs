use homotopy_core::{
    examples,
    signature::{GeneratorInfo, Signature},
    Boundary, Generator,
};
use insta::assert_debug_snapshot;

#[test]
fn assoc_left() {
    assert_debug_snapshot!(examples::associator().1.source());
}

#[test]
fn assoc_right() {
    assert_debug_snapshot!(examples::associator().1.target());
}

// ensure diagram with (source, target)
//  |
//  m     |
// / \  â¤³ |
// u  |   |
//    |
// can be created (source and target are each 3D)
#[test]
fn monoid_with_unit_to_id() -> anyhow::Result<()> {
    let (mut sig, s) = examples::scalar();
    let x = sig
        .generator_info(Generator::new(0, 0))
        .unwrap()
        .diagram()
        .clone();

    let monoid = sig.add(s.attach(&s, Boundary::Target, &[])?, s.clone())?;
    let unit = sig.add(x.identity().identity(), s.clone())?;
    let monoid_with_unit = monoid.attach(&unit, Boundary::Source, &[])?;
    assert!(sig.add(monoid_with_unit, s.identity()).is_ok());

    Ok(())
}
