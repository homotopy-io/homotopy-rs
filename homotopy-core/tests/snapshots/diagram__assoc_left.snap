---
source: homotopy-core/tests/diagram.rs
expression: "examples::associator().1.source()"

---
DiagramN {
    source: DiagramN {
        source: Diagram0(
            0:0,
        ),
        cospans: [
            GenericCospan {
                forward: Rewrite0(
                    0:0,
                    1:1,
                ),
                backward: Rewrite0(
                    0:0,
                    1:1,
                ),
            },
            GenericCospan {
                forward: Rewrite0(
                    0:0,
                    1:1,
                ),
                backward: Rewrite0(
                    0:0,
                    1:1,
                ),
            },
            GenericCospan {
                forward: Rewrite0(
                    0:0,
                    1:1,
                ),
                backward: Rewrite0(
                    0:0,
                    1:1,
                ),
            },
        ],
    },
    cospans: [
        GenericCospan {
            forward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 0,
                        internal: ConeInternal {
                            source: [
                                GenericCospan {
                                    forward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                    backward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                },
                                GenericCospan {
                                    forward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                    backward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                },
                            ],
                            target: GenericCospan {
                                forward: Rewrite0(
                                    0:0,
                                    2:2,
                                ),
                                backward: Rewrite0(
                                    0:0,
                                    2:2,
                                ),
                            },
                            slices: [
                                Rewrite0(
                                    1:1,
                                    2:2,
                                ),
                                Rewrite0(
                                    1:1,
                                    2:2,
                                ),
                            ],
                        },
                    },
                ],
                payload: (),
            },
            backward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 0,
                        internal: ConeInternal {
                            source: [
                                GenericCospan {
                                    forward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                    backward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                },
                            ],
                            target: GenericCospan {
                                forward: Rewrite0(
                                    0:0,
                                    2:2,
                                ),
                                backward: Rewrite0(
                                    0:0,
                                    2:2,
                                ),
                            },
                            slices: [
                                Rewrite0(
                                    1:1,
                                    2:2,
                                ),
                            ],
                        },
                    },
                ],
                payload: (),
            },
        },
        GenericCospan {
            forward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 0,
                        internal: ConeInternal {
                            source: [
                                GenericCospan {
                                    forward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                    backward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                },
                                GenericCospan {
                                    forward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                    backward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                },
                            ],
                            target: GenericCospan {
                                forward: Rewrite0(
                                    0:0,
                                    2:2,
                                ),
                                backward: Rewrite0(
                                    0:0,
                                    2:2,
                                ),
                            },
                            slices: [
                                Rewrite0(
                                    1:1,
                                    2:2,
                                ),
                                Rewrite0(
                                    1:1,
                                    2:2,
                                ),
                            ],
                        },
                    },
                ],
                payload: (),
            },
            backward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 0,
                        internal: ConeInternal {
                            source: [
                                GenericCospan {
                                    forward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                    backward: Rewrite0(
                                        0:0,
                                        1:1,
                                    ),
                                },
                            ],
                            target: GenericCospan {
                                forward: Rewrite0(
                                    0:0,
                                    2:2,
                                ),
                                backward: Rewrite0(
                                    0:0,
                                    2:2,
                                ),
                            },
                            slices: [
                                Rewrite0(
                                    1:1,
                                    2:2,
                                ),
                            ],
                        },
                    },
                ],
                payload: (),
            },
        },
    ],
}
